src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs

```diff
@@ -56,6 +57,25 @@ internal override void ProcessResponse(HttpResponseMessage response)
 
             if (ShouldSaveToOutFile)
             {
+                // Check if OutFile is a folder
+                if (Directory.Exists(OutFile))
+                {
+                    if (response.Headers.TryGetValues(HttpKnownHeaderNames.ContentDisposition, out IEnumerable<string> contentDisposition)) 
+                    {
+                        IEnumerator<string> enumerator = contentDisposition.GetEnumerator();
+                        if (enumerator.MoveNext())
+                        {
+                            // Get file name from Content-Disposition header if present
+                            OutFile = Path.Combine(OutFile, Path.GetFileName((string)enumerator.Current));
+                        }
+                    }
+                    else
+                    {
+                        // Get file name from last segment of Uri
+                        OutFile = Path.Combine(OutFile, System.Net.WebUtility.UrlDecode(response.RequestMessage.RequestUri.Segments[^1]));
```

Comment:
I don't value. It is better to generate an error.

Reply:
I think you're wrong, I have tested some websites and they rarely (never) use the `Content-Disposition` header. So it would always generate an error.
Tested downloads: Firefox, Chrome, 7zip, Adobe Reader

Reply:
In this case there is no point in analyzing Content-Disposition at all :-)

Reply:
Removed

Reply:
LGTM. Please update the PR description.

Reply:
Done. What should we do about the problem with `-Resume` ?

Reply:
[Content-Disposition](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) may not be used _often_ (I can't speak to that), but it is definitely used, including on GitHub:

E.g., in an archive download for a given project, such as `https://github.com/mklement0/fileicon/archive/stable.zip`, where using the last URL segment wouldn't even tell you the name of the product, the `Content-Disposition` field provides a helpful file name:

```powershell
# -> 'attachment; filename=fileicon-stable.zip'
(Invoke-WebRequest -Method Head  https://github.com/mklement0/fileicon/archive/stable.zip).Headers['Content-Disposition']
```

Both `curl` and `wget` support `Content-Disposition`, albeit on an _opt-in_ basis (`-J` and `--content-disposition`)

Reply:
@mklement0 Please describe the desired behaviour and I will implement it, I removed the check for `Content-Disposition` only because @iSazonov requested it

Reply:
Thanks, @CarloToso. 

My suggestion is as follows - though I wonder if it requires a wider discussion:

I suggest also making the feature _opt-in_, given the following warning from `man curl` with respect to its opt-in (`-J` / `--remote-header-name`):

> WARNING: Exercise judicious use of this option, especially on
          Windows. A rogue server could send you the name of a DLL or
          other file that could be loaded automatically by Windows or some
          third party software.

That requires a _new switch_, say `-UseSuggestedFileName`, which would have to report an error if the `-OutFile` argument targets a _file_.

There's also a _discoverability_ problem - also with the use-last-URI-segment approach, given that _redirections_ may be involved, and that the new default behavior uses the _redirected-to_ URI's last segment:

* How does the user learn what file name was ultimately used (short of examining the target directory for the most recently added file)?
  * At the very least, `-Verbose` should include this information.
  * `-PassThru` should use an ETS instance member to decorate the passed-through `Microsoft.PowerShell.Commands.BasicHtmlWebResponseObject` instance with a `.FullName` property that reflects the output file's full path.

As for how to parse `Content-Disposition` and fallback behavior:

* Based on the [previously linked description](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition), the C# equivalent of the following is probably good enough in practice, though the recommendation is to give precedence to `filename*` (sic) entries:

   * `<value-of-Content-Disposition-field(s)> -replace '^.*\bfilename\*?="?([^;"]+)?.*', '$1'`
   *  If there's no `Content-Disposition` field or if the file-name extraction fais, fall back to the last-URI-segment approach

Reply:
Maybe you could move this discussion to issue https://github.com/PowerShell/PowerShell/issues/11671